File: src/mvmctl/models/cloudinit.py
from enum import StrEnum, auto


class CloudInitMode(StrEnum):
    """
    Cloud-init configuration mode.

    Attributes:
        INJECT: Inject cloud-init files directly into rootfs using libguestfs (filesystem-agnostic).
        NET: Serve cloud-init files via HTTP (nocloud-net datasource).
        OFF: Skip cloud-init entirely (no ISO mounted).
        ISO: Generate cloud-init ISO from config files.

    """

    INJECT = "inject"
    NET = "net"
    OFF = "off"
    ISO = "iso"


class CloudInitStatus(StrEnum):
    """Cloud-init execution status based on console log detection."""

    PENDING = auto()
    RUNNING = auto()
    DONE = auto()
    ERROR = auto()


__all__ = ["CloudInitMode", "CloudInitStatus"]

File: internal/infra/model/cloudinit.go
package model

// ── CloudInitMode ──

// CloudInitMode represents cloud-init configuration mode.
type CloudInitMode string

const (
	CloudInitModeOFF    CloudInitMode = "off"
	CloudInitModeINJECT CloudInitMode = "inject"
	CloudInitModeNET    CloudInitMode = "net"
	CloudInitModeISO    CloudInitMode = "iso"
)

// ── CloudInitStatus ──

// CloudInitStatus represents cloud-init execution status.
type CloudInitStatus string

const (
	CloudInitStatusPending CloudInitStatus = "PENDING"
	CloudInitStatusRunning CloudInitStatus = "RUNNING"
	CloudInitStatusDone    CloudInitStatus = "DONE"
	CloudInitStatusError   CloudInitStatus = "ERROR"
)

// ── Cloud-init lifecycle interfaces ──

// CloudInitFirewallRule is a firewall rule created during cloud-init provisioning.
type CloudInitFirewallRule interface {
	Remove() error
}

// ── CloudInitResult ──

// CloudInitResult holds the result of cloud-init provisioning.
type CloudInitResult struct {
	Mode CloudInitMode

	ISOPath     *string
	NocloudURL  *string
	NocloudPort int

	NocloudPID        *int
	NocloudNetRules   []FirewallRule // Firewall rules created
}
